include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,436 @@
+  // (row, col) coordindate of the values block in a dense matrix.
+  /// </summary>
+  /// <param name="data_transfer"></param>
+  /// <param name="data_location"></param>
+  /// <param name="values_shape"></param>
+  /// <param name="values_data"></param>
+  /// <param name="indices_shape"></param>
+  /// <param name="indices_data"></param>
+  /// <returns></returns>
+  Status MakeBlockSparseData(const IDataTransfer& data_transfer,
+                             const OrtMemoryInfo& data_location,
+                             const TensorShape& values_shape, const void* values_data,
+                             const TensorShape& indices_shape, const int32_t* indices_data);
+
+  /// <summary>
+  /// Mutable data access
+  /// </summary>
+  class BlockSparseMutator {
+   public:
+    BlockSparseMutator(Tensor& values, Tensor& indices) noexcept
+        : values_(values), indices_(indices) {}
+    Tensor& Values() noexcept { return values_; }
+    Tensor& Indices() noexcept { return indices_; }
+
+   private:
+    std::reference_wrapper<Tensor> values_;
+    std::reference_wrapper<Tensor> indices_;
+  };
+
+  /// <summary>
+  /// Allocates memory for values and index and returns mutator so
+  /// data can be populated
+  /// </summary>
+  /// <param name="values_shape"></param>
+  /// <param name="indices_shape"></param>
+  /// <returns></returns>
+  BlockSparseMutator MakeBlockSparseData(const TensorShape& values_shape, const TensorShape& indices_shape);
+
+  /// <summary>
+  /// X-device copy. Destination tensor must have allocator set.
+  /// </summary>
+  /// <param name="data_transfer_manager"></param>
+  /// <param name="exec_q_id"></param>
+  /// <param name="dst_tensor"></param>
+  /// <returns></returns>
+  Status Copy(const DataTransferManager& data_transfer_manager, int exec_q_id, SparseTensor& dst_tensor) const;
+
+  /// <summary>
+  /// X-device copy. Destination tensor must have allocator set.
+  /// </summary>
+  /// <param name="dst_tensor"></param>
+  /// <returns></returns>
+  Status Copy(const IDataTransfer& data_transfer, SparseTensor& dst_tensor, int exec_q_id) const;
 
  private:
-  Tensor values_;
-  Tensor indices_;
-  TensorShape shape_;  // The shape of corresponding dense-tensor.
+  Status AllocateBuffer(int64_t buffer_size, size_t num_values);
+  void ReleaseBuffer();
+  void* IndicesStart(int64_t values_bytes);
+  const void* IndicesStart(int64_t values_bytes) const;
+  Status ValidateBlockSparseShapes(const TensorShape& values_shape, const TensorShape& index_shape) const;
+
+  std::vector<int64_t> GetCooIndexDims(size_t values_count, size_t index_size) const;
+  void InitCooIndex(const TensorShape& index_shape, const int64_t* index_data);
+
+  void ValidateCsrIndices(size_t values_count, size_t inner_size, size_t outer_size) const;
+  void InitCsrIndices(size_t inner_size, const int64_t* inner, size_t outer_size, const int64_t* outer);
```

Comment:
nit: we return Status from ValidateBlockSparseShapes but throw in ValidateCsrIndices. can we be consistent and return Status here as well (doing so gives nicer behavior in a build with no exceptions)?

Reply:
We  can't do w/o exceptions when we return mutator.